Switch to bash and make test for hacloud consistent

nicolasbock opened the pull request:
The shebang references /bin/sh which is not required to link to /bin/bash.
Since the script uses bash features (the [[ ]] test for instance) it may break
if the user decides to use a shell other than bash as /bin/sh.

The legal valus of $hacloud are specified as '' or 1 by mkcloud. This patch
makes testing of $hacloud consistent throughout the script.

aplanas commented:
For me makes sense, but this is a wasp nest. There are other 97 instances where test ( [ ] ) is used in the code. IMHO there are three options:

1.- Extend this PR until the full mkcloud code is consistent with bash
2.- Update the PR to maintain only the first line (shebang) to be explicit that we will use and expect bash
3.- Status quo

IMHO the second option is a +1 for me

@jdsn what do you think?

nicolasbock commented:
@aplanas The problem with 2 though is that a test using `test` is a lot more convoluted if we insist on testing `hacloud = { '', 1 }`. We would have to first test for `-n $hacloud` and then for `$hacloud -eq 1`. The bash way is simpler.

aspiers commented:
See also #936 - any style changes should be checked with shellcheck before merging, otherwise we risk diverging from the style it recommends by default (that might be OK, but only if we can configure it to check using a different style, and the team is mostly agreed on that style).

jdsn commented:
> 2.- Update the PR to maintain only the first line (shebang) to be explicit that we will use and expect bash
> @jdsn what do you think?

I also like to have it at least as two commits. One being the switch to bash. Another one to switch (some) checks.
Actually I don't care that much about having all checks consistent with one commit or even one PR. But it should be our goal if we do the switch.

I also had in mind such a big cleanup commit but always postponed it. So I am happy that it is brought up now.

So: +1  from me for (at least) two commits.

jdsn commented on scripts/qa_crowbarsetup.sh:
```diff
@@ -525,7 +525,7 @@ function addsp3testupdates()
     add_mount "SLES11-SP3-Updates-test" \
         $distsuseip':/dist/ibs/SUSE:/Maintenance:/Test:/SLE-SERVER:/11-SP3:/x86_64/update/' \
         "$tftpboot_repos_dir/SLES11-SP3-Updates-test/" "sp3tup"
-    [ -n "$hacloud" ] && add_mount "SLE11-HAE-SP3-Updates-test" \
```
Just for the record: this changes the semantics.

I that `['', '1']` are documented as allowed values. We had a feature using `hacloud=2`. Thats the reason why it still says `"-n $hacloud"`.
So this change will break setups where people still use `hacloud=$nonemptystring`

I am not arguing against this change, we should just be aware of this as there might be issues or questions popping up.

nicolasbock replied:
What would you think about changing that to `[0, 1, ...]`? In other words, set a default value of `0` for `hacloud`, replacing the empty string. Within this repository, the necessary changes wouldn't be all that bad. Is `hacloud` used outside this repository?

nicolasbock commented:
I started splitting the original PR into at least two. This one now only switches to bash.

jdsn commented:
> I started splitting the original PR into at least two. This one now only switches to bash.

Its still  fine to have a PR with multiple commits :)

:+1:  for this PR (even if you add the hacloud variable cleanup as separate commit)

nicolasbock commented:
Ah, good to know @jdsn :) What do you think about my proposal to change the legal values for `hacloud` to `[0, 1, ...]` and introduce a default in mkcloud?

jdsn commented:
> Ah, good to know @jdsn :) What do you think about my proposal to change the legal values for hacloud to [0, 1, ...] and introduce a default in mkcloud?

I tend to not like it that much. Because if a variable is not needed it should just not be there.
In both cases you have to check for ` = 1` anyway, so we would not gain anything (ok, in case of negation it would mean we can not rely on a ` = 0` test to be correct - just as it is right now, but this behavior is only consistent with all the `want_*` variables).

So if we switch the behavior, I'd rather recommend to rename the variable to `want_hacloud` or even `want_ha`. Then I'm fine with a default of 0 and a sanity check if the value is valid.

nicolasbock commented:
Fair enough, you have a point. The reason I changed the behavior in this way was that I didn't like that the two legal values are `''` and `1`, i.e. a string and an integer. Of course we are interpreting `1` as `"1"` and there really isn't a type difference, but from a readability point of view I thought it stood out. Would you rather me drop the second patch entirely? Or rework it so that it's more gentle?

aspiers commented:
I agree that mixing value types is aesthetically ugly, but shell is inherently ugly :-)  The only way to cleanly fix this kind of stuff is to switch to a proper language.  In the meantime I'd suggest maybe going with the path of least resistance / most consistency, whatever that is.  Certainly use of `-n` and `-z` tests are quite idiomatic in shell, and perhaps marginally more common than `= 1` or `-eq 1` etc.  But I don't have a particular strong opinion either way, other than that consistency is important.

And I'll reassert my comments about codeclimate-shellcheck, since I think that is definitely the way forward towards better shell code.

nicolasbock commented:
@aspiers Point well taken. I modified the patch a bit to make the tests around hacloud consistent.

aplanas commented:
This is an improvement: +1

nicolasbock commented:
@aspiers, can we hook shellcheck into github so that it becomes a required check? I am preparing a third commit, addressing some of the issues shellcheck found on my workstation, but it would be nice to force this check in the future.

aplanas commented:
@jdsn as a bash guru I return the assignment to you, for me this is a +1

aspiers commented:
@nicolasbock That's what I meant by codeclimate-shellcheck.  https://codeclimate.com/changelog/56e0b52837af5a008e000ba4

nicolasbock commented:
@aspiers Awesome! That looks very useful!

jdsn commented on scripts/mkcloud:
```diff
@@ -61,7 +61,7 @@ setcloudnetvars $virtualcloud
 # the default nodenumber of compute nodes
 nodenumbercomputedefault=2
-[ -n "$hacloud" ] && nodenumbercomputedefault=1
+[[ -n "$hacloud" ]] && nodenumbercomputedefault=1
```
If you cleanup do it as much as possible:
```
[[ $hacloud ]]
```

jdsn commented on scripts/mkcloud:
```diff
@@ -1256,7 +1256,7 @@ function sanity_checks()
 
     # checking clusterconfig
-    if [ -n "$hacloud" -a -z "$clusterconfig" ] ; then
+    if [[ -n "$hacloud" && -z "$clusterconfig" ]] ; then
```
```
if [[ $hacloud && ! $clusterconfig ]]
```

jdsn commented:
:+1:  + I just triggered a gating rebuild

nicolasbock commented:
@jdsn Could you have another look, please? I have made the simplification changes you suggested.

jdsn commented on scripts/qa_crowbarsetup.sh:
```diff
@@ -2656,7 +2656,7 @@ function custom_configuration()
 
             # assign neutron-network role to one of SLE12 nodes
-            if [ -n "$want_sles12" ] && [ -z "$hacloud" ] && [ -n "$want_neutronsles12" ] && iscloudver 5plus ; then
+            if [[ $want_sles12 && !$hacloud && $want_neutronsles12 && $(iscloudver 5plus) ]]; then
```
`$(iscloudver 5plus)` is 0 if its true (bash semantic: exit code 0 => ok ; exit code 1 => error)
So its wrong to move it into the `[[ ]]`.

Negating it might just work but it seems even more wrong then.
So my recommendation is to write it as:
```
if [[ $want_sles12 && !$hacloud && $want_neutronsles12 ]] &&  iscloudver 5plus ; then
```

nicolasbock replied:
Good point. I'll fix it.

jdsn commented on scripts/qa_crowbarsetup.sh:
```diff
@@ -450,11 +450,11 @@ function iscloudver()
         v=${v%M*}
         if [[ "$ver" -eq "$v" ]] && [[ $cloudsource =~ ^M[0-9]+$ ]] ; then
-            [ "${cloudsource#*M}" $operator "$milestone" ]
+            test "${cloudsource#*M}" $operator "$milestone"
```
why?

nicolasbock replied:
Following @aspiers's suggestion I ran the script through [shellcheck](https://www.shellcheck.net/) which can't handle the test expression using the `[` form, and terminates with a nonsensical error message after this line. This is a bug in `shellcheck` to be sure. Both forms are equivalent.

aspiers replied:
Hmm yeah, I would prefer that we fix `shellcheck` rather than switch to lots of `test` everywhere, which is much more old-fashioned style. Even old POSIX has `[` IIRC.

nicolasbock replied:
@jdsn and @aspiers, I changed the tests back. Maybe I'll take a look at shellcheck. I think it's written in Haskell, which should be fun :smile:

jdsn commented on scripts/qa_crowbarsetup.sh:
```diff
@@ -450,11 +450,11 @@ function iscloudver()
         fi
     fi
-    [ "$ver" $operator "$v" ]
+    test "$ver" $operator "$v"
```
why?

nicolasbock replied:
See my comment for line 453.

jdsn commented:
Still :+1:
